src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,133 @@
-""" This is a dummy example """
+import requests
+import pandas as pd
+import time
+import plotly.graph_objects as go
+from plotly.subplots import make_subplots
+
+API_URL = "https://archive-api.open-meteo.com/v1/archive?"
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+    }
+VARIABLES = ["temperature_2m_mean", "precipitation_sum", "wind_speed_10m_max"]
+
+def call_api(url:str, cool_off_time:int = 1):
+    """
+    Generic function to call API's and treats possible errors.
+    """
+
+    def rate_limit_handling():
+        """
+        Function to return a response in case of timeout error.
+        """
+        time.sleep(cool_off_time)
+        return call_api(url, cool_off_time*2) # Expontential growth of cool_off_time, arbitrary
+    
+    try:
+        response = requests.get(url)
+        if response.status_code == 200:
+            return response
+        if response.status_code == 429: # Rate limit error
+            return rate_limit_handling()
+        response.raise_for_status()
+    except requests.exceptions.Timeout as timeout_error:
+        print(f"Timeout error occurred: {timeout_error}")
+    except requests.exceptions.HTTPError as http_error: # 4XX, 5XX errors
+        print(f"HTTP error occurred: {http_error}")
+    except requests.exceptions.RequestException as request_error: # other errors
+        print(f"Request error occurred: {request_error}")
+    return None
+
+def get_data_meteo_api(city:str, start_date:str = "2010-01-01", end_date:str = "2019-12-31") -> pd.DataFrame | None:
+    """
+    Request data from Open Meteo API given a city, start_date and end_date.
+    Default start_date is "2010-01-01" and end_date is "2020-12-31".
+
+    Args:
+        city (str): City name.
+        start_date (str): Start date in format "YYYY-MM-DD".
+        end_date (str): End date in format "YYYY-MM-DD".
+    
+    Returns:
+        pd.DataFrame: Data from Open Meteo API.
+        None: In case of error.
+    """
+    latitude:float = COORDINATES[city]["latitude"]
+    longitude:float = COORDINATES[city]["longitude"]
+    url:str = f"{API_URL}latitude={latitude}&longitude={longitude}&start_date={start_date}&end_date={end_date}&daily={','.join(VARIABLES)}"
+    api_result = call_api(url)
+    if api_result is None:
+        return None
+    else:
+        data = pd.DataFrame(api_result.json()["daily"])
+        return data
+    
+def data_preprocessing(df:pd.DataFrame) -> pd.DataFrame:
```

Comment:
Name is too generic. `def compute_monthly_mean(city_data: pd.DataFrame) -> pd.DataFrame`